target/bin/setquota

```diff
@@ -15,39 +15,29 @@ USER="${1}"
 shift
 QUOTA="${*}"
 
-function usage  { echo "Usage: setquota <user@domain> [<quota>]" ; }
+function __usage { echo 'Usage: setquota <user@domain> [<quota>]' ; }
 
-[[ -z ${USER} ]] && { usage ; _exit_with_error "no username specified" ; }
-[[ ${USER} =~ .*\@.* ]] || { usage ; _exit_with_error "username must include the domain" ; }
+[[ -z ${USER} ]] && { __usage ; _exit_with_error "No username specified" ; }
+[[ ${USER} =~ .*\@.* ]] || { __usage ; _exit_with_error "Username must include the domain" ; }
 
-if ! grep -qE "^${USER}\|" "${USER_DATABASE}"
-then
-  usage; _exit_with_error "user ${USER} does not exist"
-fi
-
-# check quota
-if [[ -n ${QUOTA} ]] && ! echo "${QUOTA}" | grep -qE "^([0-9]+(B|k|M|G|T)|0)\$"
-then
-  usage
-  _exit_with_error "invalid quota format. e.g. 302M (B (byte), k (kilobyte), M (megabyte), G (gigabyte) or T (terabyte))"
-fi
+grep -qE "^${USER}\|" "${USER_DATABASE}" || _exit_with_error "User '${USER}' does not exist"
 
-_create_lock # Protect config file with lock to avoid race conditions
-
-touch "${DATABASE}"
 if [[ -z ${QUOTA} ]]
 then
-  read -r -s "Enter quota (e.g. 10M): " QUOTA
+  read -r -s 'Enter quota (e.g. 10M): ' QUOTA
   echo
-  [[ -z "${QUOTA}" ]] && _exit_with_error "Quota must not be empty. Use 0 for unlimited quota"
+  [[ -z "${QUOTA}" ]] && _exit_with_error 'Quota must not be empty (use 0 for unlimited quota)'
 fi
 
 # check quota
-if [[ -n ${QUOTA} ]] && ! echo "${QUOTA}" | grep -qE "^([0-9]+(B|k|M|G|T)|0)\$"
+if ! grep -qE "^([0-9]+(B|k|M|G|T)|0)\$" <<< "${QUOTA}"
 then
-  usage
-  _exit_with_error "invalid quota format. e.g. 302M (B (byte), k (kilobyte), M (megabyte), G (gigabyte) or T (terabyte))"
+  __usage
+  _exit_with_error 'Invalid quota format. e.g. 302M (B (byte), k (kilobyte), M (megabyte), G (gigabyte) or T (terabyte))'
 fi
 
+_create_lock # Protect config file with lock to avoid race conditions
```

Comment:
No `_remove_lock`?

Reply:
It doesn't look like a `_remove_lock` exists in the existing script.

I think this was probably because originally `flock` was used and this wasn't required. `flock` also worked properly by locking the config file being modified I believe, whereas the replacement approach only locks the script itself from being run, thus isn't as useful (eg the API project if it were to have multiple operations queried for different users or something using `delquota` for one and `setquota` for another perhaps a chance to cause problems that `flock` would avoid.

The motivation to move from `flock` was to support NFS, but when I investigated this it seemed to be a misconfiguration on the user end as `flock` supports NFS.

---

I wouldn't expect this PR to add a `_remove_lock` as a result. Correct fix is to revert to `flock`, I have plenty of notes on the subject and it's on my todo list :sweat_smile:

Reply:
The real reason `_remove_lock` is not needed is that `_create_lock` `trap`s the `EXIT` signal to call `_remove_lock` on script exit:

https://github.com/docker-mailserver/docker-mailserver/blob/f22d75fa623675f5fa7dbb05316221dd1f02b07d/target/scripts/helpers/lock.sh#L25

Reply:
I actually thought I had mentioned that in my comment, I guess I completely forgot about it when I went on my revert to `flock` rant :sweat_smile: